### ISSUE TITLE
Validation feedback/tooltips lack association with their form control

### ISSUE THREAD

**Issue by patrickhlauke** [twbs/bootstrap#28414]
the feedback and tooltip containers need to be programmatically associated with their respective form controls, e.g. using `aria-describedby="..."`, e.g.

```html
<label for="validationTooltip03">City</label>
<input type="text" class="form-control" id="validationTooltip03" ... required>
<div class="invalid-tooltip">
   Please provide a valid city.
</div>
```

to

```html
<label for="validationTooltip03">City</label>
<input type="text" class="form-control" id="validationTooltip03" aria-descriebedby="validationTooltip03-tip" ... required>
<div class="invalid-tooltip" id="validationTooltip03-tip">
   Please provide a valid city.
</div>
```

**Comment by danielnixon** [comment]
Interesting to note that this is actually a regression from Bootstrap 3 🤔 

https://getbootstrap.com/docs/3.4/css/#forms-help-text

>  Help text should be explicitly associated with the form control it relates to using the aria-describedby attribute. This will ensure that assistive technologies – such as screen readers – will announce this help text when the user focuses or enters the control.

**Comment by patrickhlauke** [comment]
sure. lots of moving parts to bootstrap, so things often fall between the cracks on large refactors/reworkings.

**Comment by patrickhlauke** [comment]
this issue is complicated by the fact that, despite `display:none`, certain browser/AT combinations will still announce the additional validation text referenced via `aria-describedby`. so these need to be added dynamically via JS, not as a pure CSS solution. thoughts @mdo ?

**Comment by patrickhlauke** [comment]
@mdo i think at this stage we probably can't rely on pure CSS to have both the "just hang off the `:valid` / `:invalid`" AND avoiding that validation tooltips get announced right away if we were properly using `aria-describedby`, as the `display:none` won't prevent them from being announced. feels like we may need an actual plugin/JavaScript behavior for form validations...thoughts? and if so, any chance of a pointer on how to go about this?

**Comment by patrickhlauke** [comment]
alternatively i guess we *could* cram all the extra logic into that example JS snippet we provide, but that feels very dirty, expecting users to manually copy/paste some custom JS in to provide actual core functionality that should be handled directly by BS silently?

**Comment by patrickhlauke** [comment]
xref https://github.com/twbs/bootstrap/pull/28995 for a first very naive approach

**Comment by patrickhlauke** [comment]
dropping this here for reference as well https://developer.paciellogroup.com/blog/2015/05/short-note-on-aria-labelledby-and-aria-describedby/

**Comment by mdo** [comment]
#31677 aims to address this in our form validation pages. Slating for alpha 3!

**Comment by XhmikosR** [comment]
@patrickhlauke @mdo: is this still valid for v5?

**Comment by patrickhlauke** [comment]
sorry, late reply to say yes, it's still a problem we need to tackle...
